row_cache.cc

```diff
@@ -51,17 +52,34 @@ row_cache::create_underlying_reader(read_context& ctx, mutation_source& src, con
 }
 
 static thread_local mutation_application_stats dummy_app_stats;
+static thread_local utils::updateable_value<double> dummy_index_cache_fraction(1.0);
 
-cache_tracker::cache_tracker(register_metrics with_metrics)
-    : cache_tracker(dummy_app_stats, with_metrics)
+cache_tracker::cache_tracker()
+    : cache_tracker(dummy_index_cache_fraction, dummy_app_stats, register_metrics::no)
+{}
+
+cache_tracker::cache_tracker(utils::updateable_value<double> index_cache_fraction, register_metrics with_metrics)
+    : cache_tracker(std::move(index_cache_fraction), dummy_app_stats, with_metrics)
 {}
 
 static thread_local cache_tracker* current_tracker;
 
-cache_tracker::cache_tracker(mutation_application_stats& app_stats, register_metrics with_metrics)
+// To implement memory partitioning between index cache and data cache,
+// we need at least a way to check how much memory is used by the index cache.
+// Currently we are relying on metrics for that.
+//
+// FIXME: relying on thread-local metrics for the algorithm is unclean.
+// It would be better to pass some accessor to these metrics as a parameter
+// to the tracker.
+namespace sstables {
+    size_t space_used_by_index_cache();
+}
+
+cache_tracker::cache_tracker(utils::updateable_value<double> index_cache_fraction, mutation_application_stats& app_stats, register_metrics with_metrics)
```

Comment:
How about passing some struct here that contains index_cache_fraction and an std::function to determine the space used by the index cache, then we avoid the dependency (which is likely circular).

Reply:
Yes, it's important to me not to reduce modularity.

Reply:
Solved by moving the ownership of index cache metrics to cache_tracker.